llmfoundry/models/hf/hf_fsdp.py

```diff
@@ -159,41 +162,31 @@ def prepare_hf_causal_lm_model_for_fsdp(model: PreTrainedModel,
-        # we need to tag all child modules that are torch.nn.Modules with `_fsdp_wrap`.
-        for child in model.children():
-            if isinstance(child, type(causal_base_model)):
-                continue
-            if isinstance(child, torch.nn.Module):
-                child._fsdp_wrap = True
-
-        for child in causal_base_model.children():
-            if isinstance(child, torch.nn.ModuleList):
-                continue
-            if isinstance(child, torch.nn.Module):
-                child._fsdp_wrap = True
-
-        if model.config.tie_word_embeddings and not model.config.model_type == 'mpt':
-            raise ValueError(
-                'The passed in HuggingFaceModel has tied word embeddings ' +
-                'and the passed in initialization device is `mixed.` ' +
-                'In order to support this initialization scheme, we would need to break '
-                +
-                'the weight tying. As a result, either use a different initialization scheme '
-                + 'or in the model config set `tie_word_embeddings=False.`')
-    else:
-        # When using the HF LM models,
-        # the weights of the self.lm_head and self.transformer.wte are tied.
-        # This tying occurs inside the `self.post_init()` function.
-        # This is a hurdle for FSDP because they need to be in the same FSDP block
-        # These lines ensures that both modules stay together in the top-most block when
-        # the model has this tying enabled (almost all do; this property defaults to True)
-        if model.config.tie_word_embeddings:
-            causal_base_model._fsdp_wrap = False
-            tied_embeddings._fsdp_wrap = False
-            lm_head._fsdp_wrap = False
+
+    # When using the HF LM models,
+    # the weights of the self.lm_head and self.transformer.wte are tied.
+    # This tying occurs inside the `self.post_init()` function.
+    # This is a hurdle for FSDP because they need to be in the same FSDP block
+    # These lines ensures that both modules stay together in the top-most block when
+    # the model has this tying enabled (almost all do; this property defaults to True)
+    if model.config.tie_word_embeddings:
+        causal_base_model._fsdp_wrap = False
+        tied_embeddings._fsdp_wrap = False
+        lm_head._fsdp_wrap = False
+
+    if hasattr(model, 'peft_type'):
+        if model.peft_type is not None:
+            peft_type = model.peft_type.lower()
+            active_adapters = [
+                adapter.lower() for adapter in model.active_adapters
+            ]
+            for name, module in model.named_modules():
+                if peft_type in name.lower() and any(
+                        adapter in name.lower() for adapter in active_adapters):
+                    has_parameters = any(True for _ in module.parameters())
+                    has_buffers = any(True for _ in module.buffers())
```

Comment:
super nit: this is a little clearer imo, but there isn't really a great solution to check if generators are empty in python
```suggestion
                    has_parameters = next(module.parameters(), None) is not None
                    has_buffers = next(module.buffers(), None) is not None
```

Reply:
i think it is also worthy of a helper function
```suggestion
                    has_parameters = _has_next(module.parameters())
                    has_buffers = _has_next(module.buffers())
```

Reply:
I don't really think this one deserves a helper since its so small and only happening right here. Can switch to your implementation though, I think they're about equal readibility and the variable name makes it pretty clear what is happening.